Lombiq.HelpfulLibraries.OrchardCore/Shapes/PerTenantShapeTableManager.cs

```diff
@@ -0,0 +1,181 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using OrchardCore.DisplayManagement.Descriptors;
+using OrchardCore.DisplayManagement.Extensions;
+using OrchardCore.Environment.Extensions;
+using OrchardCore.Environment.Extensions.Features;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Settings;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Lombiq.HelpfulLibraries.OrchardCore.Shapes;
+
+/// <summary>
+/// An altered version of <see cref="DefaultShapeTableManager"/> where the shape table is cached per tenant and per
+/// theme instead of just per theme. Also the shape descriptor collection is cached per tenant instead of being a static
+/// dictionary.
+/// </summary>
+public class PerTenantShapeTableManager : IShapeTableManager
+{
+    private readonly IHostEnvironment _hostingEnvironment;
+    private readonly IEnumerable<IShapeTableProvider> _bindingStrategies;
+    private readonly IShellFeaturesManager _shellFeaturesManager;
+    private readonly IExtensionManager _extensionManager;
+    private readonly ITypeFeatureProvider _typeFeatureProvider;
+    private readonly IMemoryCache _memoryCache;
+    private readonly ILogger _logger;
+    private readonly ISiteService _siteService;
+
+    [SuppressMessage(
+        "Major Code Smell",
+        "S107:Methods should not have too many parameters",
+        Justification = "All of these are necessary for shape table management.")]
+    public PerTenantShapeTableManager(
+        IHostEnvironment hostingEnvironment,
+        IEnumerable<IShapeTableProvider> bindingStrategies,
+        IShellFeaturesManager shellFeaturesManager,
+        IExtensionManager extensionManager,
+        ITypeFeatureProvider typeFeatureProvider,
+        IMemoryCache memoryCache,
+        ILogger<PerTenantShapeTableManager> logger,
+        ISiteService siteService)
+    {
+        _hostingEnvironment = hostingEnvironment;
+        _bindingStrategies = bindingStrategies;
+        _shellFeaturesManager = shellFeaturesManager;
+        _extensionManager = extensionManager;
+        _typeFeatureProvider = typeFeatureProvider;
+        _memoryCache = memoryCache;
+        _logger = logger;
+        _siteService = siteService;
+    }
+
+    public ShapeTable GetShapeTable(string themeId) =>
+        GetShapeTableAsync(themeId)
+            .GetAwaiter()
+            .GetResult();
+
+    public async Task<ShapeTable> GetShapeTableAsync(string themeId)
+    {
+        var siteSettings = await _siteService.GetSiteSettingsAsync();
+
+        var shapeTableCacheKey = $"ShapeTable:{siteSettings.SiteName}:{themeId}";
+        var shapeDescriptorsCacheKey = $"ShapeDescriptors:{siteSettings.SiteName}";
+
+        if (_memoryCache.TryGetValue(shapeTableCacheKey, out ShapeTable shapeTable)) return shapeTable;
```

Comment:
Isn't it better to stick with non-1liner if statements if there are already some of them? 
At least I don't like mixing 1-liners and open braces

Reply:
I always use one-liners for short-circuiting conditions if they fit.